### ISSUE TITLE
Video upload

### ISSUE THREAD

**Issue by mewlabs** [mgp25/Instagram-API#60]
Is someone figured out how to upload video? I found endpoint which should be used, but after uploading and configuring got error "media_needs_reupload, staged_position_not_found". Maybe need to upload video in several chunks, as it makes official client... Thanks in advance! @mgp25

**Comment by RageBrahimi** [comment]
Share the endpoint...

**Comment by RageBrahimi** [comment]
Share the endpoint details. Maybe this way others can try and help.

**Comment by RageBrahimi** [comment]
Share the endpoint details. Maybe this way others can try and help.

**Comment by RageBrahimi** [comment]
What was the endpoint?

**Comment by Tyhur** [comment]
+1

**Comment by tbrenci** [comment]
@mgp25 said he wouldn't be adding a feature to upload video but it should be easy to use the uploadPhoto as a template (see here: https://github.com/mgp25/Instagram-API/issues/52)

I'm currently working on the same problem in python. While it should be similar to an uploadPhoto function, I believe there are extra parameters that must be set (chunksize?) and without them it won't accept the video.

**Comment by djstanley** [comment]
Would absolutely LOVE to see this feature, whoever can do it is amazing

**Comment by beshoo** [comment]
i am ready to donate $25 for this function to be add. since it is easy as described!

**Comment by beshoo** [comment]
i am ready to donate $25 to Mr @mgp25 to make this don ! any one

**Comment by xteraflopx** [comment]
@mewlabs what version of Instagram did you test that on?

**Comment by xteraflopx** [comment]
@tbrenci I already have a working upload video function written in Python. Do you even know what you are talking about? This is not just a matter of using the "uploadPhoto" function as a template, and throwing in an extra parameter, uploading a video is an ENTIRELY separate endpoint with very specific form data that must be included in the correct way. There is no parameter called chunksize, I recommend you view the HTTP request and just copy the behaviour.

@mewlabs I know what you are doing wrong and I know what is causing the error you are receiving. Have you figured it out for yourself yet?

**Comment by mewlabs** [comment]
@xteraflopx +1, it is completelly different than uploadPhoto. Not yet, I didn't found way out and postponed this task.. Could you push on the right idea please? :)

**Comment by xteraflopx** [comment]
I don't like giving out hints because people become better reverse engineers by figuring out things for themselves. First things first, do you have a solid method of capturing the HTTP requests? That is obviously extremely important. Once you have that achieved, you need to very carefully examine the request, and make sure you mimic it completely.

**Comment by mewlabs** [comment]
@xteraflopx problem that all videos which I'm uploading from phone are splitted in several chunks, and I don't know how to do it correctly. All other things I can do for sure

**Comment by xteraflopx** [comment]
What is the method you are using to capture HTTP requests?

**Comment by mewlabs** [comment]
@xteraflopx Charles proxy

**Comment by xteraflopx** [comment]
I'm currently using a different method, but I've tried that in the past. However Charles proxy only intercepted a few endpoints... (Because certificate pinning)... I've tried Android-SSL-Trustkiller for cydia substrate, and another for the Xposed module, but both did not work with Instagram. Are you actually capturing all endpoints?

**Comment by mewlabs** [comment]
@xteraflopx yes :) I spent a lot of time to get it working... And I seems that I able to capture all, from uploading video with several "stage" chunks, uploading photo and finally configuring...

**Comment by mewlabs** [comment]
@xteraflopx I'm just didn't had the time to figure how to split video into chunks correctly, I'm think in several weeks I will have it a little to solve this puzzle :) If you want to help me with this, please write me at mewlabs gmail. And let's remove our conversation from this issue, I don't want to give this info for people who don't want to work on their own. Thanks in advance

**Comment by mgp25** [comment]
Nevermind ill do it myself this weekend, its pretty easy anyway.

**Comment by xteraflopx** [comment]
@mgp25 even @mewlabs has worked on this long enough that he doesn't want people getting free tools and info when they do no work of their own. Seriously, just let people work for it if they want it... Smart people like you and I are going to be hurt in the long run, because all the dumbos who abuse the free tools they download off Github, will draw much more attention to themselves, and Instagram will be more likely to intervene.

**Comment by mgp25** [comment]
@xteraflopx you are right, but people like us can get that easily again, so don't worry at all. I know it is a pitty that people doesnt work at their own, but is always good to share knowledge, the bad side of all of this i think you already know, its the people who use this with illicit activities and thats something im totally against with.

**Comment by mewlabs** [comment]
@mgp25 agree with @xteraflopx, need to leave some features for individual work, because this is very delicate subject

**Comment by mewlabs** [comment]
but anyway thanks @mgp25  for your position about sharing the knowledge and your overall effort :)

**Comment by MaximMikulich** [comment]
@mgp25 you see https://github.com/mgp25/Instagram-API/issues/34 ? Add location

**Comment by mgp25** [comment]
@MaximMikulich sorry im blind, i cant add it, but you can send a pull request instead

Regards

**Comment by djstanley** [comment]
Any progression with this? This code works uploading chunked video to Twitter, eg, cutting into 1mb parts and uploading one-by-one. I don't have a rooted android so I cannot sniff the endpoints, but it will be a similar process to this:

```
while (! feof($fp)) {
    $chunk = fread($fp, 1048576);
    $reply = $twit->media_upload([
      'command'       => 'APPEND',
      'media_id'      => $media_id,
      'segment_index' => $segment_id,
      'media'         => $chunk
    ]);

    $segment_id++;
}

fclose($fp);
```

**Comment by mgp25** [comment]
Im going to post Instagram video upload, the only thing i ask for you is that you need to code it and send a pull request. Thats the only thing.

I skipped some headers (because is not required to code, focus in the important part of the requests)

**fisrt request**
```
POST /api/v1/upload/video/ HTTP/1.1
Host: i.instagram.com

upload_id=1458984764906&media_type=2&_uuid=ba913fe1-5c1a-4d66-b892-b77d16c3906a&_csrftoken=d6662d21cb381d03bd55ddddd6f76aa1
```

**Response**

```json
{
	"status": "ok",
	"upload_id": "1458984764906",
	"video_upload_urls": [{
		"url": "http:\/\/upload.instagram.com\/api\/v1\/stage\/video\/?target=ash3c07\u0026vx_token=102",
		"job": "AQBjuFwCBslq6a2fBbS8ITJjgCkVmNMg_Avl-RrL4LKklJIQX7St8Uu8t8KT9N9jm-jGsjVmlkW3dkIHpNlDyPhRf8hISlLogA281xFr-VyHcM4TFWHzChaLXw1rHEOn_h6SMVRQHD7pRXfb6Lv_iMPIgrgmXmTj5WzCEx4-xtFd156ixgy6DrFwb9J19VDIACWcP-FOCI7VZMH2tGHep0t9Rxz5ykBFSZtMpZfZtkgilg5fSGwm-QDMj54PIJI5PVESHg72ts30puNXzbWI1A1qn3TThgO9oOUSjdpQHGaEN9Bp77WAoFkcdzC5NwNdlqVpI6A9plJbFnUJDyVybA5YHg-TNPQO5XAUBGCEiOZ5LavSbg38J5lKKvbKWz8nHZwFfZCCbBZb8IROzu3NJcC_VbjFZ0ur0x0c_nMKnBWEAba6QaFQNXlqDCzCuJgxu6qfoxFrOMeuVCt5r30XphOu",
		"expires": 1774344765.076154947
	}, {
		"url": "http:\/\/upload.instagram.com\/api\/v1\/stage\/video\/?target=ash3c07\u0026vx_token=100",
		"job": "AQBl0XanYUuAN_yJbhFwKTeMgWQ-f_irsrcRJM_sl--Xv7MvAX9CxaQ4KDsuajrJ8mSHyOi0ZEcfTr7Dr9doj0mEIIK0egPweioMLfVHL_8NFgF8IVlqEXxRUbXQTipnHZGAUl2HLI9_UnClD9r-1BraBAe4oiXQI7hY9NPvE2bRIGHZK-C5YY_xjIjmXrlp-mGCvOhmwJwF-CYk4NByAiQHtA2pcuU3ORo1cAR8-XUykLSyAKLBR-_hAvVQHytnFHmZjnVclz0nsLXLA14n_5qQvMn-7eBcyWf8-UVMemv81eRq7CoYV8-P1u1gxk9onXAAnkYgU_8qJagIuEWWpMR-AkCrWeW8Loz7-bpuClQ30mg4GuwpmqE5wcyT4RhNg26vhlFdZu2E6abtSRPF4eBtOCuo9Zq9LWMRA8t4hXuZ0YtinYjUb9q46a4zcFSNNqXCZ8iQViO_GGUd_kHtK5nX",
		"expires": 1774344765.077373028
	}, {
		"url": "http:\/\/upload.instagram.com\/api\/v1\/stage\/video\/?target=ash3c07\u0026vx_token=101",
		"job": "AQCJTXD2cnz2zkpPLLh61p3BNVsgvfroRzmffgCnAj82Wl60XPl89wyOJsVHqYJeEKo_HaFetNgHyHKO_XO7hMCT0fB-Od3oaUBQSDau78UugzE8BWUofw0JQs92r5QtukWXEh5Il2RJRUapsXiGcz919GbvhNv6VcZMJ96xQhFMS_bs50l7g13VPonMONZlhYhQRjziJVKxEQof9SA71rygnndw7ePT8sFsFoHieudSmB_RW3NZjll02Tfso9Aj4paVi_WxESqgWTklkuLVl8P9K0ovelUHqVEXUTaCbLOljl75namxnVeNVQLixr0-6jGsdSSxa9oZwwib0azNl6IMHDJn6o2xBMvu8FARB4fMzZx5lISmSKtaLLnJRuXjpvdnW9O7ixM-G3x33gNBCPh7m-cO72R4mecQjkH6Y48m8ijfcyVbTxICoL_qHfZpYVAjijXVW14awzx9KGVJ4ZU9",
		"expires": 1774344765.078214884
	}, {
		"url": "http:\/\/upload.instagram.com\/api\/v1\/stage\/video\/?target=ash1c16\u0026vx_token=104",
		"job": "AQA60SQaHQzpHB16ehiEszFRIcIRPprUwyASn1RKM2kmXceRODV1ZyRCK-Lnc5MZy4kn_t7oqqz-Ou9d8KlBF2vwDFCz-9WQ99i-IqfS8toihelSNmoG5-zgyyEn_M_jqGFpk_1ecNBSwplqPJVZiXW4iW6git94e0I_XYpAByzI_7tfyIeRLexG8nSplG0NsnGLBNJxV_UVSCp1uzfAmwPBr7KJta5Ye7Jb5Zm7DysHjBHYVRY57ntHeUJscSWQJ98bCKiAWmTpkgyPGrJvs3WFoeuinqGaWfQNJbXMRT4mggXWVJhnaYA7SHXoA7yz-jYrsQmKJ1uuFLMvYthBHy3U5aw9myC09ADbAyF9JD57WHN_1JIaXRBwA3VBCfqIvIKezjmrYr7Cjps9sU-K9UDWLCYE6U4jCIZeAZn3VdBsx8UQxeruG8WFb27SAtTpmOfthY4db7mjpXEI6VnRfvbj",
		"expires": 1774344765.078969955
	}]
}
```

**Video upload requests (chunks) - In this case 6 chunks**
```
POST /api/v1/stage/video/?target=ash3c07&vx_token=101 HTTP/1.1
Host: upload.instagram.com
Cookie2: $Version=1
job: AQBwFJIpMWQ91PomMs3Lbm1zqoRH_SbuaW7-HAInJBIWC_FgL1prmqeAo_EHQ5yS8PEv4NT_FCMJ-fcWEC6Tp7XiTeYqXuRvSw9yysQeMJ4X1LDDU4R3rqWbm_eveAVp9U6tcSdyKgvjChyknirKw1eRl5KHw7WsX6kQoPb0B6lbgTN4ueoZDKXul2ps5ab-0l3Pq3hOsIU6n59tz-Qwz5ppmJK8bpGFqGWGtdp3f8RasXqLqxzGGnqr6VmBal1w6IWqYiqaw-pO_VzNRxDe0MwqcwiKBK7Itm7558H4XjAkgESL88ZroYkZEmOhFY65SxKuLMXfyp01AVTvnXb9vRdU33mGr-BzHtQ-eNIaUm33PxHiw1I-rLvtGz9nFbjjGBE1jqUC22WPQins7eeO1kRKJxsk9iv_C8R076an5djCEKjBTC6dK5TjSLHD3nzcYeB0fzfZjiRQ2VUsIryTXRyg
Content-Range: bytes 0-204799/1124814
Cookie: csrftoken=152b808d6caa70c7974778d061228dac; mid=VvZXSAABAAFet8OH061XtoTpykzr

bytes
```

**Response**
```
0-204799/1124814
```

**chunk 2**

```
POST /api/v1/stage/video/?target=ash3c07&vx_token=101 HTTP/1.1
Host: upload.instagram.com
Cookie2: $Version=1
job: AQBwFJIpMWQ91PomMs3Lbm1zqoRH_SbuaW7-HAInJBIWC_FgL1prmqeAo_EHQ5yS8PEv4NT_FCMJ-fcWEC6Tp7XiTeYqXuRvSw9yysQeMJ4X1LDDU4R3rqWbm_eveAVp9U6tcSdyKgvjChyknirKw1eRl5KHw7WsX6kQoPb0B6lbgTN4ueoZDKXul2ps5ab-0l3Pq3hOsIU6n59tz-Qwz5ppmJK8bpGFqGWGtdp3f8RasXqLqxzGGnqr6VmBal1w6IWqYiqaw-pO_VzNRxDe0MwqcwiKBK7Itm7558H4XjAkgESL88ZroYkZEmOhFY65SxKuLMXfyp01AVTvnXb9vRdU33mGr-BzHtQ-eNIaUm33PxHiw1I-rLvtGz9nFbjjGBE1jqUC22WPQins7eeO1kRKJxsk9iv_C8R076an5djCEKjBTC6dK5TjSLHD3nzcYeB0fzfZjiRQ2VUsIryTXRyg
Content-Range: bytes 204800-409599/1124814
Cookie: csrftoken=152b808d6caa70c7974778d061228dac; mid=VvZXSAABAAFet8OH061XtoTpykzr

bytes
```

**response**

```
0-409599/1124814
```

(i skip some chunks and post the last one)
```
POST /api/v1/stage/video/?target=ash3c07&vx_token=101 HTTP/1.1
Host: upload.instagram.com
Cookie2: $Version=1
job: AQBwFJIpMWQ91PomMs3Lbm1zqoRH_SbuaW7-HAInJBIWC_FgL1prmqeAo_EHQ5yS8PEv4NT_FCMJ-fcWEC6Tp7XiTeYqXuRvSw9yysQeMJ4X1LDDU4R3rqWbm_eveAVp9U6tcSdyKgvjChyknirKw1eRl5KHw7WsX6kQoPb0B6lbgTN4ueoZDKXul2ps5ab-0l3Pq3hOsIU6n59tz-Qwz5ppmJK8bpGFqGWGtdp3f8RasXqLqxzGGnqr6VmBal1w6IWqYiqaw-pO_VzNRxDe0MwqcwiKBK7Itm7558H4XjAkgESL88ZroYkZEmOhFY65SxKuLMXfyp01AVTvnXb9vRdU33mGr-BzHtQ-eNIaUm33PxHiw1I-rLvtGz9nFbjjGBE1jqUC22WPQins7eeO1kRKJxsk9iv_C8R076an5djCEKjBTC6dK5TjSLHD3nzcYeB0fzfZjiRQ2VUsIryTXRyg
Content-Range: bytes 1024000-1124813/1124814
Cookie: csrftoken=152b808d6caa70c7974778d061228dac; mid=VvZXSAABAAFet8OH061XtoTpykzr

bytes
```

**Response**
```json
{
	"status": "ok",
	"result": "deprecated",
	"configure_delay_ms": "6500"
}
```

**photo upload (this is for the thumbnail)**

```
POST /api/v1/upload/photo/ HTTP/1.1
Host: i.instagram.com
Content-Type: multipart/form-data; boundary=ba913fe1-5c1a-4d66-b892-b77d16c3906a
Content-Length: 384135

--ba913fe1-5c1a-4d66-b892-b77d16c3906a
Content-Disposition: form-data; name="upload_id"

1458985465780
--ba913fe1-5c1a-4d66-b892-b77d16c3906a
Content-Disposition: form-data; name="_uuid"

ba913fe1-5c1a-4d66-b892-b77d16c3906a
--ba913fe1-5c1a-4d66-b892-b77d16c3906a
Content-Disposition: form-data; name="_csrftoken"

d6662d21cb381d03bd55ddddd6f76aa1
--ba913fe1-5c1a-4d66-b892-b77d16c3906a
Content-Disposition: form-data; name="image_compression"

{"lib_name":"jt","lib_version":"1.3.0","quality":"100"}
--ba913fe1-5c1a-4d66-b892-b77d16c3906a
Content-Disposition: form-data; name="photo"; filename="pending_media_1458985479792.jpg"
Content-Type: application/octet-stream
Content-Transfer-Encoding: binary

����JFIF����tExifMM*>F(
```

**response**

```json
{
	"status": "ok",
	"upload_id": "1458985465780"
}
```

**Media configure**

```
POST /api/v1/media/configure/?video=1 HTTP/1.1
Host: i.instagram.com

signed_body=7d4bfdafc9785b1d4607c5de35be98e4c0a0736f594a86e8131716d5c3c3fdf8.{"caption":"","upload_id":"1458985465780","source_type":"3","clips":[{"length":14.975,"source_type":"3","camera_position":"back"}],"poster_frame_index":0,"length":0.00,"audio_muted":false,"filter_type":"0","video_result":"deprecated","extra":{"source_width":960,"source_height":1280},"device":{"manufacturer":"Xiaomi","model":"HM 1SW","android_version":18,"android_release":"4.3"},"_csrftoken":"d6662d21cb381d03bd55ddddd6f76aa1","_uuid":"ba913fe1-5c1a-4d66-b892-b77d16c3906a","_uid":"2959466246"}&ig_sig_key_version=4
```

**expose**

```
POST /api/v1/qe/expose/ HTTP/1.1
Host: i.instagram.com

signed_body=dc0ec24bb95546a7cc8ddf5d86603a8b5345f8a64b39ffedc62d4d57a8e6fed4.{"_csrftoken":"d6662d21cb381d03bd55ddddd6f76aa1","experiment":"ig_android_profile_contextual_feed","id":"2959466246","_uid":"2959466246","_uuid":"ba913fe1-5c1a-4d66-b892-b77d16c3906a"}&ig_sig_key_version=4
```

**response**
```json
{
	"status": "ok"
}
```

Regards

**Comment by djstanley** [comment]
So the video is sliced into chunks, and then uploaded in chunks via the 'job' parameter?

**Comment by xteraflopx** [comment]
@djstanley the job parameter is just an ID... The place where @mgp25 wrote "bytes" is where you put the video byte data......

**Comment by mgp25** [comment]
Really? Over a week a no one has code it yet? Come on guys, don't disappoint me. I did all the stuff, are there is one step left.

**Comment by mewlabs** [comment]
@mgp25 I think everybody who need this feature, already coded it for own purposes. Other people don't want to code, they want to take ready solution and use it for spam :) jajaja

**Comment by mgp25** [comment]
Just to make things clear, if people dont collaborate with this i wont update this API anymore, so is up to you guys

**Comment by RageBrahimi** [comment]
@mgp25 I think I had the issue regarding the deprecated response part. Couldn't overcome it

**Comment by RageBrahimi** [comment]
I'm failing to receive 
```
{
    "status": "ok",
    "result": "deprecated",
    "configure_delay_ms": "6500"
}
```
I only receive 
```
string(15) "0-224359/897442"
string(20) "224360-448719/897442"
string(20) "448720-673079/897442"
string(20) "673080-897441/897442"
```

**Comment by xteraflopx** [comment]
@mewlabs you really hit the nail on the head. They don't want to code, they want to take the finished tool and use it for spam.

@RageBrahimi do you need someone to hold your hand to guide you through this? Did you even READ the complete instructions that @mgp25 gave you? You won't receive the deprecated response until you finish uploading the video.

It is so painful to see people who are unable to figure out things for themselves without asking for help (even though they were given step by step instructions)

Come on people, THINK!!!

**Comment by RageBrahimi** [comment]
@xteraflopx Listen, all you've done is moaned about people requesting help. If you've nothing good to say, I suggest you avoid visiting this thread. If you're willing to contribute, then contribute, otherwise do not aggravate people by saying they're useless. It's a community to share code, not an opportunity for you to share your disrespectful thoughts constantly.

**Comment by xteraflopx** [comment]
Hey, I did give you a tip on what you were doing wrong. And what I am TRYING to do is get people to think for themselves! You shouldn't ask for help on a forum after running into the very first issue, you should try as many solutions as possible and try to fix it yourself.

I'm not disrespecting anyone, I'm trying to provoke thought! When I wrote a video upload script for Python, I wasn't given step by step instructions from @mgp25, I actually analyzed Instagrams HTTP requests myself and figured it out. It wasn't easy, but I figured it out, so it was rewarding!

@mgp25 instructions were very clear, all I ask is that people THINK FOR THEMSELVES before posting and asking questions!

**Comment by RageBrahimi** [comment]
Actually I've been very supportive, when I can for instance [here](https://github.com/mgp25/Instagram-API/issues/67) & [here](https://github.com/mgp25/Instagram-API/issues/20) (of which you ironically didn't know the answer to) , but when I need support, I'll ask and there shouldn't be no problem in doing so and people like yourself who mock those who do. 

Rather than complaining about the questions, and claiming you've done it yourself, why don't you contribute the code and then people can learn how you've done it (by you also explaining).

I appreciate your provoking 'thought' as you say, but comments like "do you need someone to hold your hand to guide you through this" provoke the developer too, so please be more considerate.

**Comment by xteraflopx** [comment]
Okay before my rant, let me clarify how to Upload Videos.

"0-224359/897442" is the response from Instagram that confirms which bytes were successfully uploaded. Personally I upload the video in 200,000 byte chunks, and then obviously there will be a remainder to upload at the end. Once you upload all of the bytes you get the response:

```
{
    "status": "ok",
    "result": "deprecated",
    "configure_delay_ms": "6500"
}
```

Then you simply upload the photo, and then configure.

--------------------------------------------------

@RageBrahimi you've completely missed the point here.

> why don't you contribute the code and make the pull-request and then people can learn how you've done it

Ummm for the same reason @mgp25 has not coded it himself and given it away. People do not **LEARN** by looking at the end result, they learn by **DOING**. This will be my last comment on the issue so please listen carefully.

<rant>
WAY too many people post issues on this repository for things that could be solved with just a small amount of thinking. It seems that NOBODY is willing to do any work for themselves. @mgp25 and I figured out this stuff by ourselves by reverse engineering. We didn't find a repository on Github to tell us exactly what to do. In the beginning of the project it's a very daunting task. Instagram implements certificate pinning so you can't even view the HTTPS requests with Charles-Proxy. So you can either modify the .apk with ApkTool to disable certificate pinning, or you can modify the .apk to write the requests themselves to the Android Log. Then after being able to view the traffic you must understand how the signing process works. In order to sign a request you must know the secret key. How do you find the secret key? You have to use Gikdbg to look inside the binary "strings.so" and set a breakpoint, and then dump the memory to find the string.

It takes a LOT of work to do a project such as this. @mgp25 gave explicit step-by-step instructions on how Videos are upload and people STILL aren't willing to do the work. I am not trying to insult anyone, do you see me cursing anyone out? I am frustrated that people are completely unwilling to do any amount of work. It's impossible to get good at programming or reverse engineering if you ask questions the minute you run into any sign of trouble. You must go through the pain staking process of figuring our a difficult puzzle for yourself to be able to advance.
</rant>

**Comment by RageBrahimi** [comment]
Okay, there is partial progress. I've essentially replicated what @mgp25 initially provided, in a bit more detail with some supporting code.

```
POST /api/v1/upload/video/ HTTP/1.1
```
```javascript
{
	"status": "ok",
	"upload_id": "1459691185807",
	"video_upload_urls": [{
		"url": "http:\/\/upload.instagram.com\/api\/v1\/stage\/video\/?target=ash3c07\u0026vx_token=101",
		"job": "AQA9dMklARaN5xGhyIEAdwTIvFI-5KWQjPecjkqpKfJFNG_spuSpmLiTYduZPIy5zUPF9LtUTXekgZMRO-hyNCo6yOYQbQNs2RsxCtC2jh3TcPIgagPu6QBiRzD1-eCfykgc8cGZ2g2NBqG3T3YHhuZQz785JAVzaSGjg6dZKBZCVcZNwSH_CWxuaJlNTT-KesBWbYRR4ATdNJsQEPLVVFaj1AN7AF_Oh8OGe3WcY-r1XvtdWKE8C5Q0Xk2H9iyw3M9sMH-REIDvXMdGPMUEQRtbZbp2pEael-LP2pLngnyd-rBXgS6ic6jHvXqrpdeCOj6XwXm_9LJ8Gj8B6_DyIkrsuzWfwSDPOlvF1Yblrmm0K1mduzfZTy0jQZR6-6zk3B0kNhRagkhTZdOlxN5Z_iiV-3Uup9CgR_vxfOSV1l9wz6me6HDy9NHpJ3XKo4j9CQUa5et7QCYbmhpYydvVAtSb",
		"expires": 1775051186.046730995
	}, {
		"url": "http:\/\/upload.instagram.com\/api\/v1\/stage\/video\/?target=ash3c07\u0026vx_token=100",
		"job": "AQBTq2vVtnIgI830NwmKuL2mel77WZYC71rgtPqEyiBdxi1EFyHHZcH6ytZyEnFvBQThCxTsUalWoBYS5GNABbYb3MDdUIKSA8eYObeLh7iuAfwH66MxlIpZW_WNQDnLtfV7Ejzj7PjQwPua9EUNIrDgGARPoBG9jq4DsjlrPGWnrbb3a_81w1geeObtBfj7xdEAPSIGNWNfD0TMZNYG-xoiXLx0xqNJ9TJY7Wffy5GLfDyyf7-EjzsSWfde7peK-J-sRaPfrZZKvZMDUYC3MMa0StVVGqTkIQ2C572ZgLJTEzDhZBxlZ-guGprmGuoJggnbhU5u1cMaMZcIP2MMhd00cjVou7ePVj_TOXK5aNbTP71Rw-_7F5Eu9-N2pvWGR-hJ2PibivOAxvvK8ycXPUhG7iXYfUwiGNH2O3BJuil_EIQS8-nf95pmbR_WLf6Uf0Bt6tJM0-ELGLgaXYtqG5n-",
		"expires": 1775051186.047580004
	}, {
		"url": "http:\/\/upload.instagram.com\/api\/v1\/stage\/video\/?target=ash3c07\u0026vx_token=102",
		"job": "AQDt4qJ2__vPkXuQ-iBbm02eHR_9G58cZS_b7DgcfxzZkjX30dSFoogo07VL19B3_JFpa1iFXDc6mKpZgDT9eoB2-oIObfxdU7iz3ZNPukU-6Ab45exz60l44nbkCkARt-vd66YZaBHeQy2qVWjR1e8sLo2lLhajzM4wutZhCy9RxM5IKQ9osvHyjMMjO0XmkqDvr50iJdwbVtftBcytmB5HeyswPCIvHQTBti_gtRk0cNierhZF0kKIW_DwhTRwj0063jq-hLQ88sp9U5potQVHU7e_pmDMUSJ6EKkxm38eOWtLC0C_cyZza5h5wj3kL-2Qq1ADx1jKYCADMmF0dV1sJMQHn_Y0x7MhPQQmjfxg4onaZUQAev0Y_WnwrLAhfgNzTHhcHJK02SA5tXEozEZeyB83SEOMwTCUzRba4UArA27OrkEWqRow8a00aRjbgw06wa_Soga64nsRzilFJnyW",
		"expires": 1775051186.048369884
	}, {
		"url": "http:\/\/upload.instagram.com\/api\/v1\/stage\/video\/?target=ash1c16\u0026vx_token=105",
		"job": "AQAsJjiDx3_iinmMGFxiJ83JcoaHTkJfQRGzE7GDwOZllL6TfTUGQ8bsbSF18IfDLrnG8odxLnpztsZRZUNEP5CfSDCC2LWc-cRAdpeCc9HA60bCrHZpXxCXppAd8py9uR74QIMXkKXuXagG3N__MuhEgxM5DR0yUkXejC2t_G1sFGZfloRZd-wZp8R_nwyofUrZnINidMdZ_PcxADnQQDz-PRupCVPuaqs74xc8wWK2HhGa5LlY5N6rvDjyq6Ag5GESAbB6H87NWwFCbSzmb1SGwBMPzjVEovY9fEKoqgICEIAG4aIjV0GRvhstwWZswYtFS2zSwG65DvlDxiPAz32YnljrLmxVcEH6jUgtzE6wLfPEa40Wvu8qhjjwZgygwrOPYe5UnLpMphNwHx8GnpbkjCP-3avbBOmYdJpHnhIc9asMwbO1gRO1Qc8mvkUwPPGjisaoTi3Cp74gvgk1Izda",
		"expires": 1775051186.049114943
	}]
}
```

You'll get this response from using the '**uploadVideo**'  function (full function available below).
```javascript
try {
    $i->uploadVideo($video, $caption);
} catch (Exception $e) {
    echo $e->getMessage();
}
```

Dependent on the file size, you'll divide the upload process by 'chunks', in this instance, it's in **6 chunks**.

**No.1**
```
POST /api/v1/stage/video/?target=ash1c16&vx_token=105 HTTP/1.1

Host: upload.instagram.com
Cookie2: $Version=1
Content-Length: 204800
job: AQAsJjiDx3_iinmMGFxiJ83JcoaHTkJfQRGzE7GDwOZllL6TfTUGQ8bsbSF18IfDLrnG8odxLnpztsZRZUNEP5CfSDCC2LWc-cRAdpeCc9HA60bCrHZpXxCXppAd8py9uR74QIMXkKXuXagG3N__MuhEgxM5DR0yUkXejC2t_G1sFGZfloRZd-wZp8R_nwyofUrZnINidMdZ_PcxADnQQDz-PRupCVPuaqs74xc8wWK2HhGa5LlY5N6rvDjyq6Ag5GESAbB6H87NWwFCbSzmb1SGwBMPzjVEovY9fEKoqgICEIAG4aIjV0GRvhstwWZswYtFS2zSwG65DvlDxiPAz32YnljrLmxVcEH6jUgtzE6wLfPEa40Wvu8qhjjwZgygwrOPYe5UnLpMphNwHx8GnpbkjCP-3avbBOmYdJpHnhIc9asMwbO1gRO1Qc8mvkUwPPGjisaoTi3Cp74gvgk1Izda
Content-Range: bytes 0-204799/1124814
```
**No.2**
```
POST /api/v1/stage/video/?target=ash1c16&vx_token=105 HTTP/1.1

Host: upload.instagram.com
Cookie2: $Version=1
Content-Length: 204800
job: AQAsJjiDx3_iinmMGFxiJ83JcoaHTkJfQRGzE7GDwOZllL6TfTUGQ8bsbSF18IfDLrnG8odxLnpztsZRZUNEP5CfSDCC2LWc-cRAdpeCc9HA60bCrHZpXxCXppAd8py9uR74QIMXkKXuXagG3N__MuhEgxM5DR0yUkXejC2t_G1sFGZfloRZd-wZp8R_nwyofUrZnINidMdZ_PcxADnQQDz-PRupCVPuaqs74xc8wWK2HhGa5LlY5N6rvDjyq6Ag5GESAbB6H87NWwFCbSzmb1SGwBMPzjVEovY9fEKoqgICEIAG4aIjV0GRvhstwWZswYtFS2zSwG65DvlDxiPAz32YnljrLmxVcEH6jUgtzE6wLfPEa40Wvu8qhjjwZgygwrOPYe5UnLpMphNwHx8GnpbkjCP-3avbBOmYdJpHnhIc9asMwbO1gRO1Qc8mvkUwPPGjisaoTi3Cp74gvgk1Izda
Content-Range: bytes 204800-409599/1124814
```
**No.3**
```
POST /api/v1/stage/video/?target=ash1c16&vx_token=105 HTTP/1.1

Host: upload.instagram.com
Cookie2: $Version=1
Content-Length: 204800
job: AQAsJjiDx3_iinmMGFxiJ83JcoaHTkJfQRGzE7GDwOZllL6TfTUGQ8bsbSF18IfDLrnG8odxLnpztsZRZUNEP5CfSDCC2LWc-cRAdpeCc9HA60bCrHZpXxCXppAd8py9uR74QIMXkKXuXagG3N__MuhEgxM5DR0yUkXejC2t_G1sFGZfloRZd-wZp8R_nwyofUrZnINidMdZ_PcxADnQQDz-PRupCVPuaqs74xc8wWK2HhGa5LlY5N6rvDjyq6Ag5GESAbB6H87NWwFCbSzmb1SGwBMPzjVEovY9fEKoqgICEIAG4aIjV0GRvhstwWZswYtFS2zSwG65DvlDxiPAz32YnljrLmxVcEH6jUgtzE6wLfPEa40Wvu8qhjjwZgygwrOPYe5UnLpMphNwHx8GnpbkjCP-3avbBOmYdJpHnhIc9asMwbO1gRO1Qc8mvkUwPPGjisaoTi3Cp74gvgk1Izda
Content-Range: bytes 409600-614399/1124814
```
**No.4**
```
POST /api/v1/stage/video/?target=ash1c16&vx_token=105 HTTP/1.1

Host: upload.instagram.com
Cookie2: $Version=1
Content-Length: 204800
job: AQAsJjiDx3_iinmMGFxiJ83JcoaHTkJfQRGzE7GDwOZllL6TfTUGQ8bsbSF18IfDLrnG8odxLnpztsZRZUNEP5CfSDCC2LWc-cRAdpeCc9HA60bCrHZpXxCXppAd8py9uR74QIMXkKXuXagG3N__MuhEgxM5DR0yUkXejC2t_G1sFGZfloRZd-wZp8R_nwyofUrZnINidMdZ_PcxADnQQDz-PRupCVPuaqs74xc8wWK2HhGa5LlY5N6rvDjyq6Ag5GESAbB6H87NWwFCbSzmb1SGwBMPzjVEovY9fEKoqgICEIAG4aIjV0GRvhstwWZswYtFS2zSwG65DvlDxiPAz32YnljrLmxVcEH6jUgtzE6wLfPEa40Wvu8qhjjwZgygwrOPYe5UnLpMphNwHx8GnpbkjCP-3avbBOmYdJpHnhIc9asMwbO1gRO1Qc8mvkUwPPGjisaoTi3Cp74gvgk1Izda
Content-Range: bytes 614400-819199/1124814
```
**No.5**
```
POST /api/v1/stage/video/?target=ash1c16&vx_token=105 HTTP/1.1

Host: upload.instagram.com
Cookie2: $Version=1
Content-Length: 204800
job: AQAsJjiDx3_iinmMGFxiJ83JcoaHTkJfQRGzE7GDwOZllL6TfTUGQ8bsbSF18IfDLrnG8odxLnpztsZRZUNEP5CfSDCC2LWc-cRAdpeCc9HA60bCrHZpXxCXppAd8py9uR74QIMXkKXuXagG3N__MuhEgxM5DR0yUkXejC2t_G1sFGZfloRZd-wZp8R_nwyofUrZnINidMdZ_PcxADnQQDz-PRupCVPuaqs74xc8wWK2HhGa5LlY5N6rvDjyq6Ag5GESAbB6H87NWwFCbSzmb1SGwBMPzjVEovY9fEKoqgICEIAG4aIjV0GRvhstwWZswYtFS2zSwG65DvlDxiPAz32YnljrLmxVcEH6jUgtzE6wLfPEa40Wvu8qhjjwZgygwrOPYe5UnLpMphNwHx8GnpbkjCP-3avbBOmYdJpHnhIc9asMwbO1gRO1Qc8mvkUwPPGjisaoTi3Cp74gvgk1Izda
Content-Range: bytes 819200-1023999/1124814
```
**No.6** _(and final)_
```
POST /api/v1/stage/video/?target=ash1c16&vx_token=105 HTTP/1.1

Host: upload.instagram.com
Cookie2: $Version=1
Content-Length: 100814
job: AQAsJjiDx3_iinmMGFxiJ83JcoaHTkJfQRGzE7GDwOZllL6TfTUGQ8bsbSF18IfDLrnG8odxLnpztsZRZUNEP5CfSDCC2LWc-cRAdpeCc9HA60bCrHZpXxCXppAd8py9uR74QIMXkKXuXagG3N__MuhEgxM5DR0yUkXejC2t_G1sFGZfloRZd-wZp8R_nwyofUrZnINidMdZ_PcxADnQQDz-PRupCVPuaqs74xc8wWK2HhGa5LlY5N6rvDjyq6Ag5GESAbB6H87NWwFCbSzmb1SGwBMPzjVEovY9fEKoqgICEIAG4aIjV0GRvhstwWZswYtFS2zSwG65DvlDxiPAz32YnljrLmxVcEH6jUgtzE6wLfPEa40Wvu8qhjjwZgygwrOPYe5UnLpMphNwHx8GnpbkjCP-3avbBOmYdJpHnhIc9asMwbO1gRO1Qc8mvkUwPPGjisaoTi3Cp74gvgk1Izda
Content-Range: bytes 1024000-1124813/1124814
```
In the ideal world, the above will give the following response: 
```
{
	"status": "ok",
	"result": "deprecated",
	"configure_delay_ms": "6500"
}
```
**Things you'll notice**
* POST endpoint remains the same (the latest).
* JOB ID remains the same (the latest).

However, the code that I have (with great thanks to @mgp25) I'm currently not receiving that final response.

Here is the code:
```javascript
  public function uploadVideo($video, $caption = "")
  {
    if (!$this->isLoggedIn)
    {
      throw new InstagramException("Not logged in\n");
      return;
    }

    $videoData = file_get_contents($video);

    $endpoint = Constants::API_URL. 'upload/video/';
    $boundary = $this->uuid;
    $upload_id = round(microtime(true)*1000);
    $bodies = [
      [
        'type' => 'form-data',
        'name' => 'upload_id',
        'data' => $upload_id
      ],
      [
        'type' => 'form-data',
        'name' => '_uuid',
        'data' => $this->uuid
      ]
    ];

    $data = $this->buildBody($bodies,$boundary);
    $headers = [
        'Proxy-Connection: keep-alive',
        'Connection: keep-alive',
        'Accept: */*',
        'Content-type: multipart/form-data; boundary='.$boundary,
        'Accept-Language: en-en',
    ];

    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $endpoint);
    curl_setopt($ch, CURLOPT_USERAGENT, Constants::USER_AGENT);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
    curl_setopt($ch, CURLOPT_HEADER, true);
    curl_setopt($ch, CURLOPT_VERBOSE, false);
    curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($ch, CURLOPT_COOKIEFILE, $this->IGDataPath . "$this->username-cookies.dat");
    curl_setopt($ch, CURLOPT_COOKIEJAR, $this->IGDataPath . "$this->username-cookies.dat");
    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $data);

    $resp       = curl_exec($ch);
    $header_len = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
    $body       = json_decode(substr($resp, $header_len), true);
    $request_size = floor(strlen($videoData)/4);
    $lastRequestExtra = strlen($videoData) - ($request_size*4);
    $uploadUrls = $body['video_upload_urls'];

    $i = 0;
    for ($i; $i < 4; $i++)
    {
      $uploadUrl = $uploadUrls[$i]['url'];
      $job = $uploadUrls[3]['job'];
      $start = ($i*$request_size);
      $end = ($i+1)*$request_size+($i == 3?$lastRequestExtra:0);
      $headers = [
          'Connection: keep-alive',
          'Accept: */*',
          'Cookie2: $Version=1',
          'Accept-Encoding: gzip, deflate',
          'Content-Type: application/octet-stream',
          'Session-ID: ' . $upload_id,
          'Accept-Language: en-en',
          'Content-Disposition: attachment; filename="video.mov"',
          'Content-Length: ' . ($end-$start),
          'Content-Range: ' . "bytes ".$start."-" .($end-1)."/" . strlen($videoData),
          'job' => $job
      ];

      $ch = curl_init();
      curl_setopt($ch, CURLOPT_URL, $uploadUrl);
      curl_setopt($ch, CURLOPT_USERAGENT, Constants::USER_AGENT);
      curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
      curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
      curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
      curl_setopt($ch, CURLOPT_HEADER, true);
      curl_setopt($ch, CURLOPT_VERBOSE, false);
      curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
      curl_setopt($ch, CURLOPT_COOKIEFILE, $this->IGDataPath . "$this->username-cookies.dat");
      curl_setopt($ch, CURLOPT_COOKIEJAR, $this->IGDataPath . "$this->username-cookies.dat");
      curl_setopt($ch, CURLOPT_POST, true);
      curl_setopt($ch, CURLOPT_POSTFIELDS, substr($videoData,$start,$end));

      $result = curl_exec($ch);
      $header_len = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
      $body       = substr($result, $header_len);
      var_dump($body);
    }

  }
```
**Things you'll notice**
* POST endpoint doesn't remains the same.
* JOB ID remains the same (the latest).
* Uploaded in 4 chunks.

If you do find a fix, **please post the amended code here** and I will finish off the rest. (please don't be selfish by not posting it if you do find out how). Evidently @mgp25 wants to see how we work together so lets show we can.

**Comment by mgp25** [comment]
@RageBrahimi :+1: you are on the way

and for the rest, calm all tits there is no reason for hate, and start struggling with the code. Explanation why im not giving the code, and instead giving you instructions:

Someday maybe i'll get tired of updating and doing things like this, so people should collaborate together to get a global aim. Remeber that taking the first step can be harder than the rest of the challenge. Once you are there trying and failing, over and over again until you succeed, that's the journey.

This way im forcing or im trying to force people to think by theirself, to learn and give them the ability to do more code and updates in their own and this way collaborating.

Remember this: If you never stop trying, you will get there eventually.

Have a nice Monday

Regards

**Comment by wehmoen** [comment]
Just a random thougth:

In the last step it is

Content-Range: bytes 1024000-1124813/1124814

Maybe Instagram expects the last missing byte?

**Comment by mgp25** [comment]
@wehmoen count 10. There are many ways to count 10:

1, 2, 3, 4, 5, 6, 7, 8, 9 and 10

You can also do:

0, 1, 2, 3, 4, 5, 6, 7, 8, and 9

In both cases the length remains the same. If you go to the first chunk:

`Content-Range: bytes 0-204799/1124814`

It starts in zero :wink:

**Comment by wehmoen** [comment]
😅 Ups... Hm okay. No more ideas atm. But I won't give up ._. It must be possible to solve this last problem :smile:

**Comment by Inviseble** [comment]
`curl_setopt($ch, CURLOPT_POSTFIELDS, substr($videoData,$start,$end));`
> function substr ($string, $start, $length = null)
$end can $request_size ?

**Comment by Inviseble** [comment]
if use
`$uploadUrl = $uploadUrls[$i]['url'];`
i get 
`
string '0-204799/1523859' (length=16)
string '204800-409599/1523859' (length=21)
string '409600-614399/1523859' (length=21)
string '614400-819199/1523859' (length=21)
.......
`
good work return 
`
0-204799/1124814
0-409599/1124814
.......
`

If select rand element and use only it and splite file on 204800
i get result
`
string '0-204799/1523859' (length=16)
string '0-409599/1523859' (length=16)
string '0-614399/1523859' (length=16)
string '0-819199/1523859' (length=16)
string '0-1023999/1523859' (length=17)
string '0-1228799/1523859' (length=17)
string '0-1433599/1523859' (length=17)
string 'Invalid job received from client' (length=32)
`
param => job
is one in all request

i dont cnow where problem :-((

**Comment by Inviseble** [comment]
i see other steps in snifer...
1) upload/video/ - get urls
2) upload photo
3) send configure for video
4) send video
---------------------
I will try tomorrow

**Comment by Arei15** [comment]
Hi, I try get send configure for video, but I get server response 400 Bad Request. Whats  my problem?

**Comment by RageBrahimi** [comment]
@arei15 have you resolved my issue above before you tried to send the configure for video?

**Comment by Arei15** [comment]
Mayby, I writer by C#  and I uploaded video and photo, but I don't know how configure video.

**Comment by RageBrahimi** [comment]
Convert your C# to PHP and post it here (so you offer a fix) and I said, once someone gives me a fix, I will write the code to configure the video upload.

**Comment by Arei15** [comment]
I don't know php. I can send C# code . Can you please write HTTP header from configure the video

**Comment by RageBrahimi** [comment]
Upload the supposed fix you have for C# on here and maybe we can look into it and then someone can hopefully convert it.
Once converted I'll do the media configure for video.

**Comment by Arei15** [comment]
I upload video in Instagram...

**Comment by RageBrahimi** [comment]
You should explain how, support others.

**Comment by Arei15** [comment]
Ofcouse, but later

**Comment by RageBrahimi** [comment]
@Arei15 Well, the sooner the better so we can finally get this issues resolved.

**Comment by wehmoen** [comment]
@Arei15 hey :) can you post your c# code?

**Comment by Arei15** [comment]
Hi, I can write my C# code tomorrow, and my code working without "expose", it's normal?

**Comment by RageBrahimi** [comment]
Write it today, so we can get this over and done with ASAP.

**Comment by xteraflopx** [comment]
@ragebrahimi don't tell him what to do, maybe this issue would have been solved "asap" if you wouldn't have given up and waited for someone to come along to give you the completed code.

**Comment by Arei15** [comment]
@mgp25 

    public string UploadVideo(byte[] previewData, byte[] videoData, string caption)
        {
            string upload_id = ((long)(DateTime.UtcNow - new DateTime(1970, 1, 1, 0, 0, 0)).TotalSeconds).ToString();

            string uuid = Guid.NewGuid().ToString();

            #region Upload video

            {
                Uri uri = new Uri(HttpUrl + "upload/video/");

                HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
                request.Method = "POST";
                request.UserAgent = UserAgent;
                request.Host = "i.instagram.com";
                request.UseDefaultCredentials = true;
                request.AllowAutoRedirect = false;
                request.KeepAlive = true;
                request.CookieContainer = new CookieContainer();
                request.CookieContainer.Add(uri, _cookies);

                byte[] upload_video_data_array = Encoding.UTF8.GetBytes($"upload_id={upload_id}&media_type=2&_uuid={uuid}&_csrftoken={_csrftoken}");

                using (Stream stream = request.GetRequestStream())
                {
                    stream.Write(upload_video_data_array, 0, upload_video_data_array.Length);
                }

                try
                {
                    using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
                    {
                        using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                        {
                            dynamic node = new JNode(reader.ReadToEnd());

                            if (node != null)
                            {
                                if (node.ContainsKey("video_upload_urls"))
                                {
                                    JList urls = node.video_upload_urls;

                                    if (urls != null)
                                    {
                                        if (urls.Count > 0)
                                        {
                                            string url = urls[3].url.Replace("\\u0026", "&").Replace("\\", "");

                                            string job = urls[3].job;

                                            if (!string.IsNullOrEmpty(url) && !string.IsNullOrEmpty(job))
                                            {
                                                long chunk = Convert.ToInt64(Math.Floor(videoData.Length / 4.0));

                                                long last = (videoData.Length - (chunk * 4));

                                                List<byte> data = videoData.ToList();

                                                for (int i = 0; i < 4; i++)
                                                {
                                                    long start = (i * chunk);

                                                    long end = ((i + 1) * chunk) + ((i == 3) ? last : 0);

                                                    #region Upload chunk

                                                    Uri chunk_uri = new Uri(url);
                                                    HttpWebRequest chunk_request = (HttpWebRequest)WebRequest.Create(chunk_uri);
                                                    chunk_request.Method = "POST";
                                                    chunk_request.Host = "upload.instagram.com";
                                                    chunk_request.Accept = "*/*";
                                                    chunk_request.Headers.Add("Cookie2", "$Version=1");
                                                    chunk_request.ContentType = "application/octet-stream";
                                                    chunk_request.ContentLength = end - start;
                                                    chunk_request.Headers.Add("Session-ID", upload_id);
                                                    chunk_request.Headers.Add("Content-Disposition", "attachment; filename=\"2.mov\"");
                                                    chunk_request.UserAgent = UserAgent;
                                                    chunk_request.UseDefaultCredentials = true;
                                                    chunk_request.Headers.Add("Content-Range", $"bytes {start}-{end - 1}/{videoData.Length}");
                                                    chunk_request.Headers.Add("job", job);
                                                    chunk_request.AllowAutoRedirect = false;
                                                    chunk_request.CookieContainer = new CookieContainer();
                                                    chunk_request.CookieContainer.Add(chunk_uri, _cookies);
                                                    chunk_request.KeepAlive = true;

                                                    byte[] chunk_data = data.GetRange((int)start, (int)(end - start)).ToArray();

                                                    using (Stream chunk_stream = chunk_request.GetRequestStream())
                                                    {
                                                        chunk_stream.Write(chunk_data, 0, chunk_data.Length);
                                                    }

                                                    HttpWebResponse chunk_response = (HttpWebResponse)chunk_request.GetResponse();

                                                    chunk_response.Close();

                                                    #endregion
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
                catch
                {
                    return null;
                }
            }

            #endregion

            #region Upload preview

            {
                Uri uri = new Uri(HttpUrl + "upload/photo/");
                HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
                request.Method = "POST";
                request.UserAgent = UserAgent;
                request.Host = "i.instagram.com";
                request.Accept = "*/*";
                request.ContentType = $"multipart/form-data; boundary={uuid}";
                request.Headers.Add("Cookie2", "$Version=1");
                request.UseDefaultCredentials = true;
                request.AllowAutoRedirect = false;
                request.CookieContainer = new CookieContainer();
                request.CookieContainer.Add(uri, _cookies);
                request.KeepAlive = true;
                
                string data = string.Empty;

                data += $"\r\n\r\n--{uuid}\r\n";
                data += $"Content-Disposition: form-data; name=\"upload_id\"\r\n";
                data += $"\r\n";
                data += $"{upload_id}\r\n";
                data += $"--{uuid}\r\n";
                data += $"Content-Disposition: form-data; name=\"_uuid\"\r\n";
                data += $"\r\n";
                data += $"{uuid}\r\n";
                data += $"--{uuid}\r\n";
                data += $"Content-Disposition: form-data; name=\"_csrftoken\"\r\n";
                data += $"\r\n";
                data += $"{_csrftoken}\r\n";
                data += $"--{uuid}\r\n";
                data += $"Content-Disposition: form-data; name=\"image_compression\"\r\n";
                data += $"\r\n";
                data += "{\"lib_name\":\"jt\",\"lib_version\":\"1.3.0\",\"quality\":\"100\"}\r\n";
                data += $"--{uuid}\r\n";
                data += $"Content-Disposition: form-data; name=\"photo\"; filename=\"pending_media_{upload_id}.jpg\"\r\n";
                data += $"Content-Type: application/octet-stream\r\n";
                data += $"Content-Transfer-Encoding: binary\r\n\r\n";

                byte[] begin_block = Encoding.UTF8.GetBytes(data);

                byte[] end_block = Encoding.UTF8.GetBytes($"\r\n--{uuid}--");
                
                request.ContentLength = (begin_block.Length + previewData.Length + end_block.Length);

                try
                {
                    using (var stream = request.GetRequestStream())
                    {
                        stream.Write(begin_block, 0, begin_block.Length);
                        stream.Write(previewData, 0, previewData.Length);
                        stream.Write(end_block, 0, end_block.Length);
                    }
                }
                catch 
                {
                    return null;
                }
                
                try
                {
                    HttpWebResponse response = (HttpWebResponse)request.GetResponse();

                    response.Close();
                }
                catch
                {
                    return null;
                }
            }

            #endregion

            #region Configure video

            {
                string configure = $"{{\"caption\":\"{caption}\",\"upload_id\":\"{upload_id}\",\"source_type\":\"3\", \"camera_position\":\"unknown\",\"extra\":{{\"source_width\":1280,\"source_height\":720}},\"clips\":[{{\"length\":10.0,\"creation_date\" :\"2016-04-09T19:03:32-0700\",\"source_type\":\"3\",\"camera_position\":\"back\"}}],\"poster_frame_index\":0,\"audio_muted\":false,\"filter_type\":\"0\",\"video_result\":\"deprecated\",\"_csrftoken\":\"{_csrftoken}\",\"_uuid\":\"{uuid}\",\"_uid\":\"{UserId}\"}}";
 
                string signature = GetSignature(configure);

                byte[] signed = Encoding.UTF8.GetBytes("signed_body=" + Uri.EscapeDataString(signature + "." + configure) + "&ig_sig_key_version=4");

                Uri uri = new Uri(HttpsUrl + "media/configure/?video=1");

                HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
                request.Method = "POST";
                request.UserAgent = UserAgent;
                request.Host = "i.instagram.com";
                request.Accept = "*/*";
                request.CookieContainer = new CookieContainer();
                request.CookieContainer.Add(uri, _cookies);
                request.ContentLength = signed.Length;

                try
                {
                    using (Stream stream = request.GetRequestStream())
                    {
                        stream.Write(signed, 0, signed.Length);
                    }
                }
                catch
                {
                    return null;
                }

                try
                {
                    using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
                    {
                        using (Stream stream = response.GetResponseStream())
                        {
                            using (StreamReader reader = new StreamReader(stream))
                            {
                                return reader.ReadToEnd();
                            }
                        }
                    }
                }
                catch (WebException e)
                {
                    //var resp = new StreamReader(e.Response.GetResponseStream()).ReadToEnd();

                    return null;
                }
            }

            #endregion
        }

**Comment by Arei15** [comment]
If you have qustion, write)))

**Comment by RageBrahimi** [comment]
Hi @Arei15, appreciate you posting the C# code. 
Could you please tell me the things you've additionally added from the code I've posted.

Also, when I use `$uploadUrl = $uploadUrls[3]['url'];`, I get
```
string(16) "0-519748/2078998"
string(17) "0-1039497/2078998"
string(17) "0-1559246/2078998"
string(32) "Invalid job received from client"
```
How did you fix this in your code?

**Comment by infoquestsolutions** [comment]
@Arei15 Are you sure your code is working because now it says your app need to update to upload photo seems like new IG_SIG_KEY value required .

**Comment by Arei15** [comment]
You must return 

string(15) "0-224359/897442"
string(20) "224360-448719/897442"
string(20) "448720-673079/897442"
string(20) "673080-897441/897442"

but don't

string(16) "0-519748/2078998"
string(17) "0-1039497/2078998"
string(17) "0-1559246/2078998"
string(32) "Invalid job received from client"

You have incorrect fierst element.

**Comment by RageBrahimi** [comment]
@Arei15 When I use `$uploadUrl = $uploadUrls[$i]['url'];` I only get the following response:
```
string(15) "0-224359/897442"
string(20) "224360-448719/897442"
string(20) "448720-673079/897442"
string(20) "673080-897441/897442"
```
But, I don't get the additional required information:
```javascript
{
    "status": "ok",
    "result": "deprecated",
    "configure_delay_ms": "6500"
}
```
However, if I try `$uploadUrl = $uploadUrls[3]['url'];` I get the following response:
```
string(16) "0-519748/2078998"
string(17) "0-1039497/2078998"
string(17) "0-1559246/2078998"
string(32) "Invalid job received from client"
```

Do you understand my issue?

**Comment by dubmartin** [comment]
@xteraflopx 

> don't tell him what to do, maybe this issue would have been solved "asap" if you wouldn't have given up and waited for someone to come along to give you the completed code.

Why don't you wind it in? If you have nothing constructive to say, just don't say anything.

You were not always so wise as you make it. I read another thread on this project in the early days where you posted basic questions and got help / information. 

I will give you the benefit of the doubt, as perhaps English is not your native language and you are unaware of how you come across.



--------------------------------
Anyway, @RageBrahimi  have you watched your headers in a nework sniffer so you can compare with what you are sending and what the app is sending?

The error :""Invalid job received from client" is implying the job ID you are sending is wrong.

**Comment by RageBrahimi** [comment]
@dubmartin 
I've not used any no. With regards to the Job ID, on the examples above:
```javascript
POST /api/v1/stage/video/?target=ash1c16&vx_token=105 HTTP/1.1

Host: upload.instagram.com
Cookie2: $Version=1
Content-Length: 204800
job: AQAsJjiDx3_iinmMGFxiJ83JcoaHTkJfQRGzE7GDwOZllL6TfTUGQ8bsbSF18IfDLrnG8odxLnpztsZRZUNEP5CfSDCC2LWc-cRAdpeCc9HA60bCrHZpXxCXppAd8py9uR74QIMXkKXuXagG3N__MuhEgxM5DR0yUkXejC2t_G1sFGZfloRZd-wZp8R_nwyofUrZnINidMdZ_PcxADnQQDz-PRupCVPuaqs74xc8wWK2HhGa5LlY5N6rvDjyq6Ag5GESAbB6H87NWwFCbSzmb1SGwBMPzjVEovY9fEKoqgICEIAG4aIjV0GRvhstwWZswYtFS2zSwG65DvlDxiPAz32YnljrLmxVcEH6jUgtzE6wLfPEa40Wvu8qhjjwZgygwrOPYe5UnLpMphNwHx8GnpbkjCP-3avbBOmYdJpHnhIc9asMwbO1gRO1Qc8mvkUwPPGjisaoTi3Cp74gvgk1Izda
Content-Range: bytes 0-204799/1124814
```
And all the other chunks, the Job ID remains the same...

**Comment by mewlabs** [comment]
guys, try to upload file without splitting into chunks. It works for me, but I didn't used this library ever, and still don't have time to convert my raw code into method for this library... If you can not succeed, in coming days I'll try to make PR

**Comment by dubmartin** [comment]
@RageBrahimi 
Do check out your headers, must probably the reason why it fails is staring at you (usually is in my case :)

**Comment by 454447415244** [comment]
I'm not using this library, I just found it on Google. With my own code, I'm getting "416 Requested Range Not Satisfiable" at the end. Any clue why? Below is the dump for my requests.

Initial Request
Server Reply => {"status":"ok","upload_id":"1461379407820","video_upload_urls":[{"url":"http:\/\/upload.instagram.com\/api\/v1\/stage\/video\/?target=prn1c22\u0026vx_token=104","job":"AQCx0EHVG3H6vzxUB4DMJ9UBQ... etc]}

Content-Range: bytes 0-204799/1217713
Server Reply 0-204799/1217713

Content-Range: bytes 204800-409599/1217713
Server Reply 0-409599/1217713

Content-Range: bytes 409600-614399/1217713
Server Reply => 0-614399/1217713

Content-Range: bytes 614400-819199/1217713
Server Reply => 0-819199/1217713

Content-Range: bytes 819200-1023999/1217713
Server Reply => 0-1023999/1217713

Content-Range: bytes 1024000-1217712/1217713
Server Reply => <html>
<head><title>416 Requested Range Not Satisfiable</title></head>
<body bgcolor="white">
<center><h1>416 Requested Range Not Satisfiable</h1></center>
<hr><center>nginx</center>
</body>
</html>

**Comment by 454447415244** [comment]
I don't get it why my comment got deleted. I wanted to share my code once it is done.

**Comment by xteraflopx** [comment]
Completely off topic comment:

@mgp25 did Google seriously donate like $50 dollars to you in BTC?
![mgp25](https://cloud.githubusercontent.com/assets/16569582/14860150/6522f320-0c75-11e6-8f94-e49ceb45051d.png)

That's pretty awesome :)

**Comment by wehmoen** [comment]
@454447415244 may you create a pull request please? :)

**Comment by Kruizi** [comment]
@RageBrahimi 
![screenshot_5](https://cloud.githubusercontent.com/assets/11752811/14902188/36b5e302-0dbb-11e6-9d0c-2947d6e07e09.png)

Returns, then what you need

        $videoData = file_get_contents($video);

        $endpoint = Constants::API_URL. 'upload/video/';
        $boundary = $this->uuid;
        $upload_id = round(microtime(true)*1000);
        $bodies = [
            [
                'type' => 'form-data',
                'name' => 'upload_id',
                'data' => $upload_id
            ],
            [
                'type' => 'form-data',
                'name' => '_csrftoken',
                'data' => $this->token,
            ],
            [
                'type' => 'form-data',
                'name' => 'media_type',
                'data'   => '2',
            ],
            [
                'type' => 'form-data',
                'name' => '_uuid',
                'data' => $this->uuid
            ]
        ];

        $data = $this->buildBody($bodies,$boundary);
        $headers = [
            'Connection: keep-alive',
            'Accept: */*',
            'Host: i.instagram.com',
            'Content-type: multipart/form-data; boundary='.$boundary,
            'Accept-Language: en-en',
        ];

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $endpoint);
        curl_setopt($ch, CURLOPT_USERAGENT, Constants::USER_AGENT);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
        curl_setopt($ch, CURLOPT_HEADER, true);
        curl_setopt($ch, CURLOPT_VERBOSE, false);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_COOKIEFILE, $this->IGDataPath . "$this->username-cookies.dat");
        curl_setopt($ch, CURLOPT_COOKIEJAR, $this->IGDataPath . "$this->username-cookies.dat");
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, $data);

        $resp       = curl_exec($ch);
        $header_len = curl_getinfo($ch, CURLINFO_HEADER_SIZE);

        $body       = json_decode(substr($resp, $header_len), true);


        $uploadUrl = $body['video_upload_urls'][3]['url'];
        $job = $body['video_upload_urls'][3]['job'];

        $request_size = floor(strlen($videoData)/4);

        $lastRequestExtra = (strlen($videoData) - ($request_size*4));

        for($a = 0; $a <= 3; $a++){
            $start = ($a*$request_size);
            $end = ($a+1)*$request_size+($a == 3?$lastRequestExtra:0);

            $headers = [
                'Connection: keep-alive',
                'Accept: */*',
                'Host: upload.instagram.com',
                'Cookie2: $Version=1',
                'Accept-Encoding: gzip, deflate',
                'Content-Type: application/octet-stream',
                'Session-ID: ' . $upload_id,
                'Accept-Language: en-en',
                'Content-Disposition: attachment; filename="video.mov"',
                'Content-Length: ' . ($end-$start),
                'Content-Range: ' . "bytes ".$start."-" .($end-1)."/" . strlen($videoData),
                'job: '.$job
            ];
            $ch = curl_init();
            curl_setopt($ch, CURLOPT_URL, $uploadUrl);
            curl_setopt($ch, CURLOPT_USERAGENT, Constants::USER_AGENT);
            curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
            curl_setopt($ch, CURLOPT_HEADER, true);
            curl_setopt($ch, CURLOPT_VERBOSE, false);
            curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
            curl_setopt($ch, CURLOPT_COOKIEFILE, $this->IGDataPath . "$this->username-cookies.dat");
            curl_setopt($ch, CURLOPT_COOKIEJAR, $this->IGDataPath . "$this->username-cookies.dat");
            curl_setopt($ch, CURLOPT_POST, true);
            curl_setopt($ch, CURLOPT_POSTFIELDS, substr($videoData,$start,$end));

            $result = curl_exec($ch);
            $header_len = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
            $body       = substr($result, $header_len);
            $array[] = [$body];
        }

**Comment by RageBrahimi** [comment]
Thanks @Kruizi I'll post the rest soon and make a pull-request shortly after.

**Comment by Kruizi** [comment]
@RageBrahimi 
Good luck to you

**Comment by 454447415244** [comment]
@Kruizi Any specific reason why you chosed video_upload_urls[3] and not 0, 1 or 2?

**Comment by mgp25** [comment]
sigh... https://github.com/mgp25/Instagram-API/commit/1b52f8eb1ef6f96e8a73b2323ac8b5694d73bbd7

requires ffmpeg

Regards

**Comment by MaximMikulich** [comment]
@mgp25  sigh...can be an example of how to download videos from the thumbnail?

**Comment by haiduong87** [comment]
@Arei15
Can you help me the upload video code?
I'm able to upload video, upload preview.
But when I try to configure the video, the server told me to upgrade instagram

**Comment by Arsintorabi** [comment]
i get this err:
code
 $result = $Instagram->uploadVideo($video_path, "hi insta");
res:
array(2) { ["status"]=> string(4) "fail" ["message"]=> string(39) "Other media configure error: yEmZkUpAj4" }

**Comment by makahleh** [comment]
@Arei15  
Hi ! 
I copied your code after i had read all the comments.
The thing is now for the first request (/upload/video) it always return error "Internal Server Error".
I think the problem somewhere inupload_video_data_array,
I tried almost everything i could but with no luck.

Can you help me please?
@mgp25
